src/libstd/time/duration.rs

```diff
@@ -168,15 +173,114 @@ impl Div<u32> for Duration {
 
 impl fmt::Display for Duration {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        match (self.secs, self.nanos) {
-            (s, 0) => write!(f, "{}s", s),
-            (0, n) if n % NANOS_PER_MILLI == 0 => write!(f, "{}ms",
-                                                         n / NANOS_PER_MILLI),
-            (0, n) if n % 1_000 == 0 => write!(f, "{}µs", n / 1_000),
-            (0, n) => write!(f, "{}ns", n),
-            (s, n) => write!(f, "{}.{}s", s,
-                             format!("{:09}", n).trim_right_matches('0'))
+        // FIXME: fmt::Formatter::with_padding() isn't public, and the pad() function applies
+        // precision in a way we don't want. At such time as it becomes reasonable to support
+        // padding without reimplementing fmt::Formatter::with_padding() here, this should be
+        // updated to support padding. In preparation for that, we're already writing to a
+        // stack buffer.
+
+        // µ (U+00B5) is 0xC2B5 in UTF-8.
+        // The following is >= the longest possible string we can format.
+        let mut buf = *b"18446744073709551615.999999999\xC2\xB5s";
+
+        fn write_to_buf<'a>(dur: &Duration, f: &fmt::Formatter, mut buf: &'a mut [u8])
```

Comment:
I don't think using a buffer allocation would simplify it, that would just end up writing to a `Vec<u8>` instead of a `Cursor<&mut [u8]>`, but the code otherwise would be the same. The complexity here comes just from the fact that I'm writing to an intermediate buffer at all, which I want to preserve because I think we should support padding/alignment on all stdlib `fmt::Display` implementations. The `fmt` API doesn't currently expose the necessary tools to do this, but if nobody else deals with that, I'm planning on at some point trying to figure out the right way to fix that so we can support padding properly.

Some of the complexity also comes from the fact that the `flt2dec` code is all private. I'd love to be able to explicitly call into that code with integral values for the various components, but for now I'm using the hack documented in this code. I figure I'll take a look again in the future when our new float formatting code is in place.

I do agree that this is a bit complicated. It ended up being more complicated than I had first expected. Part of that is due to the need to round the decimal when not printing at full precision (this rounding is the main reason I fell back to using the float serializing).